File: articles/trait-hierarchies-facets-and-the-architecture-of-personality/r/analyze_trait_hierarchies.R
args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
script_path <- if (length(file_arg) > 0) normalizePath(sub("^--file=", "", file_arg[1])) else normalizePath("r/analyze_trait_hierarchies.R")
root <- normalizePath(file.path(dirname(script_path), ".."))
data_path <- file.path(root, "data", "synthetic_hierarchical_trait_items.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

domains <- c("extraversion_score", "agreeableness_score", "conscientiousness_score", "neuroticism_score", "openness_score")
facets <- c("sociability_score", "assertiveness_score", "compassion_score", "politeness_score", "orderliness_score", "industriousness_score", "anxiety_score", "volatility_score", "aesthetics_score", "intellect_score")
outcomes <- c("broad_life_functioning", "focused_reliability_outcome", "creative_engagement_outcome")
derived <- c("bandwidth_fidelity_gap", "facet_profile_dispersion", "hierarchy_consistency_index")

alpha_score <- function(df) {
  k <- ncol(df)
  item_vars <- apply(df, 2, var)
  total_var <- var(rowSums(df))
  (k / (k - 1)) * (1 - sum(item_vars) / total_var)
}

data$conscientiousness_rescored <- rowMeans(data[, c("c1", "c2", "c3", "c4", "c5", "c6")])
data$orderliness_rescored <- rowMeans(data[, c("o1", "o2", "o3")])
data$industriousness_rescored <- rowMeans(data[, c("i1", "i2", "i3")])

reliability <- data.frame(
  scale = c("conscientiousness_c1_c6", "orderliness_o1_o3", "industriousness_i1_i3", "broad_item_pool_item1_item60"),
  level = c("domain_item_set", "facet_item_set", "facet_item_set", "broad_item_pool"),
  n_items = c(6, 3, 3, 60),
  cronbach_alpha = c(
    alpha_score(data[, c("c1", "c2", "c3", "c4", "c5", "c6")]),
    alpha_score(data[, c("o1", "o2", "o3")]),
    alpha_score(data[, c("i1", "i2", "i3")]),
    alpha_score(data[, paste0("item", 1:60)])
  )
)

domain_model <- lm(broad_life_functioning ~ extraversion_score + agreeableness_score + conscientiousness_score + neuroticism_score + openness_score, data = data)
focused_domain_model <- lm(focused_reliability_outcome ~ conscientiousness_score, data = data)
focused_facet_model <- lm(focused_reliability_outcome ~ orderliness_score + industriousness_score, data = data)
creative_domain_model <- lm(creative_engagement_outcome ~ openness_score, data = data)
creative_facet_model <- lm(creative_engagement_outcome ~ aesthetics_score + intellect_score, data = data)

models <- list(
  broad_life_functioning_from_domains = domain_model,
  focused_reliability_from_domain = focused_domain_model,
  focused_reliability_from_facets = focused_facet_model,
  creative_engagement_from_domain = creative_domain_model,
  creative_engagement_from_facets = creative_facet_model
)

model_fit <- data.frame(
  model = names(models),
  r_squared = sapply(models, function(m) summary(m)$r.squared),
  adj_r_squared = sapply(models, function(m) summary(m)$adj.r.squared),
  n = nrow(data)
)

coef_table <- function(model, model_name) {
  coefs <- summary(model)$coefficients
  data.frame(model = model_name, term = rownames(coefs), estimate = coefs[,1], standard_error = coefs[,2], t_value = coefs[,3], p_value = coefs[,4], row.names = NULL)
}

model_coefficients <- do.call(rbind, Map(coef_table, models, names(models)))

write.csv(summary(data[, c(domains, outcomes, derived)]), file.path(outputs, "r_domain_summary.csv"))
write.csv(summary(data[, c(facets, outcomes, derived)]), file.path(outputs, "r_facet_summary.csv"))
write.csv(cor(data[, c(domains, facets, outcomes, derived)], use = "pairwise.complete.obs"), file.path(outputs, "r_trait_hierarchy_correlations.csv"))
write.csv(reliability, file.path(outputs, "r_reliability_summary.csv"), row.names = FALSE)
write.csv(model_fit, file.path(outputs, "r_model_fit_summary.csv"), row.names = FALSE)
write.csv(model_coefficients, file.path(outputs, "r_model_coefficients.csv"), row.names = FALSE)
write.csv(data, file.path(outputs, "r_scored_hierarchical_trait_items.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")

File: articles/trait-hierarchies-facets-and-the-architecture-of-personality/python/analyze_trait_hierarchies.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_hierarchical_trait_items.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

DOMAINS = ["extraversion_score", "agreeableness_score", "conscientiousness_score", "neuroticism_score", "openness_score"]
FACETS = [
    "sociability_score", "assertiveness_score", "compassion_score", "politeness_score",
    "orderliness_score", "industriousness_score", "anxiety_score", "volatility_score",
    "aesthetics_score", "intellect_score"
]
OUTCOMES = ["broad_life_functioning", "focused_reliability_outcome", "creative_engagement_outcome"]
DERIVED = ["bandwidth_fidelity_gap", "facet_profile_dispersion", "hierarchy_consistency_index"]

REQUIRED = {"respondent_id", *[f"item{i}" for i in range(1, 61)], "c1", "c2", "c3", "c4", "c5", "c6", "o1", "o2", "o3", "i1", "i2", "i3", *DOMAINS, *FACETS, *OUTCOMES, *DERIVED}

def cronbach_alpha(frame: pd.DataFrame) -> float:
    clean = frame.dropna()
    n_items = clean.shape[1]
    if n_items <= 1:
        return float("nan")
    item_variances = clean.var(axis=0, ddof=1)
    total_score = clean.sum(axis=1)
    total_var = total_score.var(ddof=1)
    return float((n_items / (n_items - 1)) * (1 - item_variances.sum() / total_var)) if total_var else float("nan")

def ols(df: pd.DataFrame, y: str, xs: list[str]) -> dict:
    model_df = df[[y] + xs].dropna()
    X = np.column_stack([np.ones(len(model_df)), model_df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, model_df[y].to_numpy(float), rcond=None)
    pred = X @ beta
    ss_res = float(np.sum((model_df[y].to_numpy(float) - pred) ** 2))
    ss_tot = float(np.sum((model_df[y].to_numpy(float) - model_df[y].mean()) ** 2))
    r2 = 1.0 - ss_res / ss_tot if ss_tot else float("nan")
    return {"coefficients": dict(zip(["intercept", *xs], map(float, beta))), "r_squared": r2, "n": int(len(model_df))}

df = pd.read_csv(DATA)
missing = REQUIRED - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

# Re-score key domain and facet examples from item columns.
df["conscientiousness_rescored"] = df[["c1", "c2", "c3", "c4", "c5", "c6"]].mean(axis=1)
df["orderliness_rescored"] = df[["o1", "o2", "o3"]].mean(axis=1)
df["industriousness_rescored"] = df[["i1", "i2", "i3"]].mean(axis=1)

reliability = pd.DataFrame([
    {"scale": "conscientiousness_c1_c6", "level": "domain_item_set", "n_items": 6, "cronbach_alpha": cronbach_alpha(df[["c1", "c2", "c3", "c4", "c5", "c6"]])},
    {"scale": "orderliness_o1_o3", "level": "facet_item_set", "n_items": 3, "cronbach_alpha": cronbach_alpha(df[["o1", "o2", "o3"]])},
    {"scale": "industriousness_i1_i3", "level": "facet_item_set", "n_items": 3, "cronbach_alpha": cronbach_alpha(df[["i1", "i2", "i3"]])},
    {"scale": "broad_item_pool_item1_item60", "level": "broad_item_pool", "n_items": 60, "cronbach_alpha": cronbach_alpha(df[[f"item{i}" for i in range(1, 61)]])},
])

domain_summary = df[DOMAINS + OUTCOMES + DERIVED].agg(["mean", "std", "min", "max"]).T.reset_index().rename(columns={"index": "variable"})
facet_summary = df[FACETS + OUTCOMES + DERIVED].agg(["mean", "std", "min", "max"]).T.reset_index().rename(columns={"index": "variable"})
correlations = df[DOMAINS + FACETS + OUTCOMES + DERIVED].corr()

models = {
    "broad_life_functioning_from_domains": ols(df, "broad_life_functioning", DOMAINS),
    "focused_reliability_from_conscientiousness_domain": ols(df, "focused_reliability_outcome", ["conscientiousness_score"]),
    "focused_reliability_from_conscientiousness_facets": ols(df, "focused_reliability_outcome", ["orderliness_score", "industriousness_score"]),
    "creative_engagement_from_openness_domain": ols(df, "creative_engagement_outcome", ["openness_score"]),
    "creative_engagement_from_openness_facets": ols(df, "creative_engagement_outcome", ["aesthetics_score", "intellect_score"]),
    "bandwidth_fidelity_gap_from_profile_dispersion": ols(df, "bandwidth_fidelity_gap", ["facet_profile_dispersion", "hierarchy_consistency_index"]),
    "professional_use_boundary": "Suitable for education, research prototyping, psychometric demonstration, consulting support, organizational learning, coaching reflection, and methodological comparison; not a standalone assessment, diagnosis, placement, screening, prediction, or decision system.",
}

model_fit = pd.DataFrame([
    {"model": name, "r_squared": spec["r_squared"], "n": spec["n"]}
    for name, spec in models.items()
    if isinstance(spec, dict)
])

domain_summary.to_csv(OUT / "python_domain_summary.csv", index=False)
facet_summary.to_csv(OUT / "python_facet_summary.csv", index=False)
reliability.to_csv(OUT / "python_reliability_summary.csv", index=False)
correlations.to_csv(OUT / "python_trait_hierarchy_correlations.csv")
model_fit.to_csv(OUT / "python_model_fit_summary.csv", index=False)
(OUT / "python_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_hierarchical_trait_items.csv", index=False)
print(f"Wrote outputs to {OUT}")
